/**:
  ros__parameters:
    log_to_file: true
    log_debug: true
    log_enabled:
      #- navigation
      - navigation.graph_map_server
      - navigation.command
        # tactic
      - tactic
      - tactic.pipeline
      - tactic.module
      - tactic.module.live_mem_manager
      - tactic.module.graph_mem_manager
      - torch
        # path planner
        # "path_planning",
        # "path_planning.teb",
        # "path_planning.cbit",
        # "path_planning.cbit_planner",
        # "mpc.cbit",
        # "mpc_debug.cbit",
        # mission planner
      - mission.server
      - mission.state_machine
        # pose graph
      - pose_graph
        # pipeline specific
        # "lidar.pipeline",
        # "lidar.honeycomb_converter",
        # "lidar.preprocessing",
        # "lidar.odometry_icp",
        # "lidar.odometry_map_maintenance",
        # "lidar.vertex_test",
        # "lidar.localization_map_recall",
        # "lidar.localization_icp",
        # "lidar.intra_exp_merging",
        # "lidar.dynamic_detection",
        # "lidar.inter_exp_merging",
        # "lidar.change_detection",
        # "lidar.ground_extraction",
        # "lidar.obstacle_detection",
      # "lidar.terrain_assessment",
    robot_frame: base_link
    env_info_topic: env_info
    lidar_frame: honeycomb
    lidar_topic: /points
    graph_map:
      origin_lat: 43.7822
      origin_lng: -79.4661
      origin_theta: 1.3
      scale: 1.0
    tactic:
      enable_parallelization: true
      preprocessing_skippable: false
      odometry_mapping_skippable: false
      localization_skippable: false
      task_queue_num_threads: 1
      task_queue_size: -1
      route_completion_translation_threshold: 2.0
      chain:
        min_cusp_distance: 1.5
        angle_weight: 7.0
        search_depth: 5
        search_back_depth: 10
        distance_warning: 5.0
      save_odometry_result: true
      save_localization_result: true
      visualize: true
      rviz_loc_path_offset:
        - 0.0
        - 0.0
        - 0.0
    pipeline:
      type: lidar
      preprocessing:
        - conversion
        - filtering
      odometry:
        - icp
        - mapping
        - vertex_test
        - intra_exp_merging
        - dynamic_detection
        - inter_exp_merging
        - memory
        - network
      localization:
        - recall
        - icp
        - safe_corridor
        - change_detection_sync
        - memory
      submap_translation_threshold: 1.5
      submap_rotation_threshold: 30.0
    preprocessing:
      conversion:
        type: lidar.honeycomb_converter_v2
        visualize: true
      filtering:
        type: lidar.preprocessing
        num_threads: 8
        crop_range: 40.0
        frame_voxel_size: 0.2
        vertical_angle_res: 0.0132645
        polar_r_scale: 2.0
        r_scale: 4.0
        h_scale: 1.54
        num_sample1: 10000
        min_norm_score1: 0.95
        num_sample2: 10000
        min_norm_score2: 0.2
        min_normal_estimate_dist: 1.0
        max_normal_estimate_angle: 0.44
        cluster_num_sample: 10000
        visualize: true
    odometry:
      network:
        type: torch.sample
        use_gpu: true
        abs_filepath: false
        filepath: sample_model.pt
      icp:
        type: lidar.odometry_icp
        use_trajectory_estimation: true
        traj_num_extra_states: 0
        traj_lock_prev_pose: false
        traj_lock_prev_vel: false
        traj_qc_diag:
          - 1.0
          - 0.1
          - 0.1
          - 0.1
          - 0.1
          - 1.0
        num_threads: 8
        first_num_steps: 2
        initial_max_iter: 4
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3
        averaging_num_steps: 2
        verbose: false
        max_iterations: 1
        min_matched_ratio: 0.5
        visualize: true
      mapping:
        type: lidar.odometry_map_maintenance_v2
        map_voxel_size: 0.2
        crop_range_front: 40.0
        back_over_front_ratio: 0.5
        point_life_time: 20.0
        visualize: true
      vertex_test:
        type: lidar.vertex_test
        max_translation: 0.3
        max_rotation: 10.0
      intra_exp_merging:
        type: lidar.intra_exp_merging_v2
        depth: 6.0
        map_voxel_size: 0.2
        crop_range_front: 40.0
        back_over_front_ratio: 0.5
        visualize: true
      dynamic_detection:
        type: lidar.dynamic_detection
        depth: 12.0
        horizontal_resolution: 0.041
        vertical_resolution: 0.026
        max_num_observations: 2000
        min_num_observations: 4
        dynamic_threshold: 0.3
        visualize: true
      inter_exp_merging:
        type: lidar.inter_exp_merging_v2
        map_voxel_size: 0.2
        max_num_experiences: 128
        distance_threshold: 0.6
        planar_threshold: 0.2
        normal_threshold: 0.8
        dynamic_obs_threshold: 1
        visualize: true
      memory:
        type: live_mem_manager
        window_size: 5
    localization:
      recall:
        type: lidar.localization_map_recall
        map_version: pointmap
        visualize: true
      icp:
        type: lidar.localization_icp
        use_pose_prior: true
        num_threads: 8
        first_num_steps: 2
        initial_max_iter: 4
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3
        averaging_num_steps: 2
        verbose: false
        max_iterations: 1
        min_matched_ratio: 0.3
      safe_corridor:
        type: lidar.safe_corridor
        lookahead_distance: 5.0
        corridor_width: 3.5
        influence_distance: 1.0
        resolution: 0.25
        size_x: 16.0
        size_y: 8.0
        visualize: true
      change_detection_sync:
        type: lidar.change_detection_v3
        detection_range: 8.0
        search_radius: 0.25
        negprob_threshold: 0.1
        use_prior: true
        alpha0: 3.0
        beta0: 0.03
        use_support_filtering: true
        support_radius: 0.25
        support_variance: 0.05
        support_threshold: 2.5
        influence_distance: 0.55
        minimum_distance: 0.35
        resolution: 0.25
        size_x: 16.0
        size_y: 8.0
        visualize: true
      change_detection:
        type: lidar.change_detection_v2
        detection_range: 8
        search_radius: 0.25
        negprob_threshold: 0.1
        use_prior: true
        alpha0: 3.0
        beta0: 0.03
        use_support_filtering: true
        support_radius: 0.25
        support_variance: 0.05
        support_threshold: 2.5
        resolution: 0.5
        size_x: 16.0
        size_y: 8.0
        run_online: false
        run_async: true
        visualize: false
        save_module_result: false
      memory:
        type: graph_mem_manager
        vertex_life_span: 5
        window_size: 3
      obstacle_detection:
        type: lidar.obstacle_detection
        z_min: 0.5
        z_max: 2.0
        resolution: 0.6
        size_x: 40.0
        size_y: 20.0
        run_async: true
        visualize: true
      ground_extraction:
        type: lidar.ground_extraction
        z_offset: 0.2
        alpha: 0.035
        tolerance: 0.25
        Tm: 0.3
        Tm_small: 0.1
        Tb: 0.5
        Trmse: 0.1
        Tdprev: 1.0
        rmin: 2.0
        num_bins_small: 30.0
        bin_size_small: 0.5
        num_bins_large: 10.0
        bin_size_large: 1.0
        resolution: 0.6
        size_x: 40.0
        size_y: 20.0
        run_async: true
        visualize: true
      terrain_assessment:
        type: lidar.terrain_assessment
        lookahead_distance: 15.0
        corridor_width: 1.0
        search_radius: 1.0
        resolution: 0.5
        size_x: 40.0
        size_y: 20.0
        run_online: false
        run_async: true
        visualize: true
    path_planning:
      type: cbit
      control_period: 33
      teb:
        visualize: true
        extrapolate: false
        extrapolation_timeout: 2.0
        lookahead_distance: 8.5
        robot_model: circular
        robot_radius: 0.5
        map_frame: planning frame
        enable_homotopy_class_planning: true
        free_goal_vel: true
        max_vel_x: 0.6
        max_vel_y: 0.0
        max_vel_theta: 0.3
        min_turning_radius: 3
        weight_viapoint: 0.5
        weight_costmap: 100.0
      cbit:
        obs_padding: 0.0
        curv_to_euclid_discretization: 20
        sliding_window_width: 10.0
        sliding_window_freespace_padding: 0.5
        corridor_resolution: 0.05
        state_update_freq: 0.5
        update_state: true
        rand_seed: 1
        initial_samples: 250
        batch_samples: 100
        pre_seed_resolution: 0.5
        alpha: 0.5
        q_max: 2.5
        frame_interval: 50
        iter_max: 10000000
        eta: 1.1
        rad_m_exhange: 1.0
        initial_exp_rad: 0.75
        extrapolation: false
        incremental_plotting: false
        plotting: true
      costmap:
        costmap_filter_value: 0.01
        costmap_history: 15
      mpc:
        horizon_steps: 40
        horizon_step_size: 0.1
        forward_vel: 1.0
        max_lin_vel: 1.25
        max_ang_vel: 0.75
        vehicle_model: unicycle
        pose_error_cov:
          - 0.5
          - 0.5
          - 0.5
          - 10.0
          - 10.0
          - 10.0
        vel_error_cov:
          - 0.1
          - 1.0
        acc_error_cov:
          - 0.1
          - 0.75
        kin_error_cov:
          - 0.001
          - 0.001
          - 0.001
          - 0.001
          - 0.001
          - 0.001
        command_history_length: 100
